webapi/src/main/scala/org/knora/webapi/slice/admin/domain/service/ProjectExportService.scala

```diff
@@ -0,0 +1,209 @@
+import org.knora.webapi.messages.util.rdf.RdfStreamProcessor
+import org.knora.webapi.messages.util.rdf.Statement
+import org.knora.webapi.messages.util.rdf.TriG
+import org.knora.webapi.slice.admin.AdminConstants.adminDataGraph
+import org.knora.webapi.slice.admin.AdminConstants.permissionsDataGraph
+import org.knora.webapi.slice.admin.domain.model.KnoraProject
+import org.knora.webapi.slice.ontology.domain.service.OntologyRepo
+import org.knora.webapi.store.triplestore.api.TriplestoreService
+
+@accessible
+trait ProjectExportService {
+
+  /**
+   * Exports a project to a file.
+   * The file format is TriG.
+   * The data exported is:
+   *  * the project metadata
+   *  * the project's permission data
+   *  * the triples of the project's ontologies
+   *
+   * @param project the project to be exported
+   * @param file the path to the file to which the project should be exported
+   * @return the [[Path]] to the file to which the project was exported
+   */
+  def exportProjectTriples(project: KnoraProject, file: Path): Task[Path]
+}
+
+/**
+ * Represents a named graph to be saved to a TriG file.
+ *
+ * @param graphIri the IRI of the named graph.
+ * @param tempDir  the directory in which the file is to be saved.
+ */
+private case class NamedGraphTrigFile(graphIri: IRI, tempDir: Path) {
+  lazy val dataFile: Path = {
+    val filename = graphIri.replaceAll("[.:/]", "_") + ".trig"
+    tempDir.resolve(filename)
+  }
+}
+
+/**
+ * An [[RdfStreamProcessor]] for combining several named graphs into one.
+ *
+ * @param formattingStreamProcessor an [[RdfStreamProcessor]] for writing the combined result.
+ */
+private class CombiningRdfProcessor(formattingStreamProcessor: RdfStreamProcessor) extends RdfStreamProcessor {
+  private var startedStatements = false
+
+  // Ignore this, since it will be done before the first file is written.
+  override def start(): Unit = {}
+
+  // Ignore this, since it will be done after the last file is written.
+  override def finish(): Unit = {}
+
+  override def processNamespace(prefix: IRI, namespace: IRI): Unit =
+    // Only accept namespaces from the first graph, to prevent conflicts.
+    if (!startedStatements) {
+      formattingStreamProcessor.processNamespace(prefix, namespace)
+    }
+
+  override def processStatement(statement: Statement): Unit = {
+    startedStatements = true
+    formattingStreamProcessor.processStatement(statement)
+  }
+}
+final case class ProjectExportServiceLive(
+  ontologyRepo: OntologyRepo,
+  projectRepo: ProjectADMService,
+  stringFormatter: StringFormatter,
+  triplestoreService: TriplestoreService,
+  messageRelay: MessageRelay
+) extends ProjectExportService {
+
+  /**
+   * Combines several TriG files into one.
+   *
+   * @param namedGraphTrigFiles the TriG files to combine.
+   * @param resultFile          the output file.
+   */
+  private def combineGraphs(namedGraphTrigFiles: Seq[NamedGraphTrigFile], resultFile: Path): Task[Unit] = ZIO.attempt {
```

Comment:
to me it feels as if this should live somewhere in the RDF util

Reply:
Good idea, maybe even a dedicated `TriG` service. I might pick that up in a later PR.